ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,842 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Component that removes the "dead" data from the set of provided log segments and reclaims space.
+ * <p/>
+ * It does this by copying the valid data into swap spaces and atomically replacing the segments being compacted with
+ * the swap spaces.
+ */
+class BlobStoreCompactor {
+  static final String INDEX_SEGMENT_READ_JOB_NAME = "blob_store_compactor_index_segment_read";
+  static final String LOG_SEGMENT_COPY_JOB_NAME = "blob_store_compactor_log_segment_copy";
+  static final String TARGET_INDEX_CLEAN_SHUTDOWN_FILE_NAME = "compactor_clean_shutdown";
+  static final FilenameFilter TEMP_LOG_SEGMENTS_FILTER = new FilenameFilter() {
+    private final String SUFFIX = LogSegmentNameHelper.SUFFIX + TEMP_LOG_SEGMENT_NAME_SUFFIX;
+
+    @Override
+    public boolean accept(File dir, String name) {
+      return name.endsWith(SUFFIX);
+    }
+  };
+
+  private static final long WAIT_TIME_FOR_CLEANUP_MS = 5 * Time.MsPerSec;
+  private static final String TEMP_LOG_SEGMENT_NAME_SUFFIX = BlobStore.SEPARATOR + "temp";
+
+  private final File dataDir;
+  private final String storeId;
+  private final StoreKeyFactory storeKeyFactory;
+  private final StoreConfig config;
+  private final StoreMetrics metrics;
+  private final Log srcLog;
+  private final DiskIOScheduler diskIOScheduler;
+  private final ScheduledExecutorService scheduler;
+  private final MessageStoreRecovery recovery;
+  private final Time time;
+  private final UUID sessionId;
+  private final UUID incarnationId;
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  private volatile boolean isActive = false;
+  private PersistentIndex srcIndex;
+
+  private Log tgtLog;
+  private PersistentIndex tgtIndex;
+  private long numSwapsUsed;
+  private StoreFindToken recoveryStartToken = null;
+  private CompactionLog compactionLog;
+  private volatile CountDownLatch runningLatch = new CountDownLatch(0);
+
+  /**
+   * Constructs the compactor component.
+   * @param dataDir the directory where all compactions need to be run.
+   * @param storeId the unique ID of the store.
+   * @param storeKeyFactory the {@link StoreKeyFactory} to generate {@link StoreKey} instances.
+   * @param config the {@link StoreConfig} that defines configuration parameters.
+   * @param metrics the {@link StoreMetrics} to use to record metrics.
+   * @param diskIOScheduler the {@link DiskIOScheduler} to schedule I/O.
+   * @param srcLog the {@link Log} to copy data from.
+   * @param scheduler the {@link ScheduledExecutorService} to use.
+   * @param recovery the {@link MessageStoreRecovery} to use to recover the index and log.
+   * @param time the {@link Time} instance to use.
+   * @param sessionId the sessionID of the store.
+   * @param incarnationId the incarnation ID of the store.
+   */
+  BlobStoreCompactor(String dataDir, String storeId, StoreKeyFactory storeKeyFactory, StoreConfig config,
```

Comment:
Currently, there seem to be no characteristically/fundamentally different designs for executing the compaction so I do not see any gains with trying to go down the interface route. It is also hard to design APIs for an interface when it is not clear what the common denominator of different implementations could be.

Further, the API set of the `BlobStoreCompactor` is small and simple enough that if we were to find a reason to introduce an interface in the future, it will be easy to do so.